### PR TITLE
[SPARK-11663][Streaming]Add Java API for trackStateByKey

### PR DESCRIPTION
**Pull request by zsxwing**
TODO
- [x] Add Java API
- [x] Add API tests
- [x] Add a function test

### CODE REVIEW

File: examples/src/main/java/org/apache/spark/examples/streaming/JavaStatefulNetworkWordCount.java [hunk 1]
```diff
@@ -105,9 +87,22 @@ public static void main(String[] args) {
           }
         });
 
+    // Update the cumulative count function
+    final Function4<Time, String, Optional<Integer>, JavaState<Integer>, Optional<Tuple2<String, Integer>>> trackStateFunc =
```
Comment (tdas) [inline examples/src/main/java/org/apache/spark/examples/streaming/JavaStatefulNetworkWordCount.java]: Can you use the simple function example?

Comment (zsxwing) [inline examples/src/main/java/org/apache/spark/examples/streaming/JavaStatefulNetworkWordCount.java]: Need the key to emit the word and its count

**Comment by zsxwing** [comment]
Removed JavaState and JavaStateSpec

File: streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala [hunk 2]
```diff
@@ -92,10 +91,8 @@ sealed abstract class StateSpec[KeyType, ValueType, StateType, EmittedType] exte
 /**
  * :: Experimental ::
  * Builder object for creating instances of [[org.apache.spark.streaming.StateSpec StateSpec]]
- * that is used for specifying the parameters of the DStream transformation
- * `trackStateByKey` operation of a
- * [[org.apache.spark.streaming.dstream.PairDStreamFunctions pair DStream]] (Scala) or a
- * [[org.apache.spark.streaming.api.java.JavaPairDStream JavaPairDStream]] (Java).
+ * that is used for specifying the parameters of the DStream transformation `trackStateByKey`
+ * operation of a [[org.apache.spark.streaming.dstream.PairDStreamFunctions pair DStream]].
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala]: Please add JavaPairDStream back in the docs.

File: streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala [hunk 3]
```diff
@@ -123,8 +119,8 @@ object StateSpec {
   /**
    * Create a [[org.apache.spark.streaming.StateSpec StateSpec]] for setting all the specifications
    * `trackStateByKey` operation on a
-   * [[org.apache.spark.streaming.dstream.PairDStreamFunctions pair DStream]] (Scala) or a
-   * [[org.apache.spark.streaming.api.java.JavaPairDStream JavaPairDStream]] (Java).
+   * [[org.apache.spark.streaming.dstream.PairDStreamFunctions pair DStream]].
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala]: Add java back in.

File: streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala [hunk 4]
```diff
@@ -142,8 +138,8 @@ object StateSpec {
...
+   *
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala]: Add java back in.

File: streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala (hunk 3 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala]: Nvm, I got it.

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaTrackStateDStream.scala [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.api.java.JavaSparkContext
+import org.apache.spark.streaming.dstream.TrackStateDStream
+
+/**
+ * :: Experimental ::
+ * [[JavaDStream]] representing the stream of records emitted by the tracking function in the
+ * `trackStateByKey` operation on a [[JavaPairDStream]]. Additionally, it also gives access to the
+ * stream of state snapshots, that is, the state data of ll keys after a batch has updated them.
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaTrackStateDStream.scala]: ll --> all

File: streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala [hunk 6]
```diff
@@ -123,8 +119,8 @@ object StateSpec {
   /**
    * Create a [[org.apache.spark.streaming.StateSpec StateSpec]] for setting all the specifications
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala]: `specifications of the ..`

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaTrackStateDStream.scala [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+package org.apache.spark.streaming.api.java
+
+import org.apache.spark.annotation.Experimental
+import org.apache.spark.api.java.JavaSparkContext
+import org.apache.spark.streaming.dstream.TrackStateDStream
+
+/**
+ * :: Experimental ::
+ * [[JavaDStream]] representing the stream of records emitted by the tracking function in the
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaTrackStateDStream.scala]: Does this automatically link to the Scala docs of JavaDStream? May be safer to do `[[org.apache....JavaDStream JavaDStream]]`

File: core/src/main/java/org/apache/spark/api/java/function/Function4.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+ */
+
+package org.apache.spark.api.java.function;
+
+import java.io.Serializable;
+
+/**
+ * A four-argument function that takes arguments of type T1, T2, T3 and T4 and returns an R.
+ */
+public interface Function4<T1, T2, T3, T4, R> extends Serializable {
```
Comment (tdas) [inline core/src/main/java/org/apache/spark/api/java/function/Function4.java]: Can you make sure you add a test for this in Java8?
they are in extras/java-8-tests/...

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala [hunk 9]
```diff
@@ -426,6 +428,48 @@ class JavaPairDStream[K, V](val dstream: DStream[(K, V)])(
...
+   *   Function2<Optional<Integer>, State<Integer>, Optional<String>> trackStateFunc =
+   *       new Function2<Optional<Integer>, State<Integer>, Optional<String>>() {
+   *
+   *         @Override
+   *         public Optional<String> call(Optional<Integer> one, State<Integer> state) {
+   *           // Check if state exists, accordingly update/remove state and return transformed data
+   *         }
+   *       };
+   *
+   *    JavaTrackStateDStream[Integer, Integer, Integer, String] trackStateDStream =
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala]: Shouldnt these be `JavaTrackStateDStream<Integer, Integer, Integer, String>`??

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala]: Make sure that the code would compile

Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaPairDStream.scala]: Just went through all examples and fixed them

**Comment by zsxwing** [comment]
retest this please

File: core/src/main/java/org/apache/spark/api/java/function/Function4.java (hunk 8 above)
Comment (zsxwing) [inline core/src/main/java/org/apache/spark/api/java/function/Function4.java]: Done

File: streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala (hunk 6 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/StateSpec.scala]: Done

File: streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaTrackStateDStream.scala (hunk 7 above)
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/api/java/JavaTrackStateDStream.scala]: I think it's fine for classes in the same package. Take a look at: http://spark.apache.org/docs/latest/api/scala/index.html#org.apache.spark.sql.DataFrame

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by tdas** [comment]
Thanks @zsxwing I am merging this to master and 1.6

**Pull request closed without merging**